pkg/codegen/docs/gen.go

```diff
@@ -1753,61 +1757,201 @@ func (mod *modContext) getModuleFileName() string {
 	return mod.mod
 }
 
+// moduleConflictResolver holds module-level information for resolving naming conflicts.
+// It shares information with the top-level docGenContext
+// to ensure the same name is used across modules that reference each other.
+type moduleConflictResolver struct {
+	dctx *docGenContext
+	seen map[string]struct{}
+}
+
+func (dctx *docGenContext) newModuleConflictResolver() moduleConflictResolver {
+	return moduleConflictResolver{
+		dctx: dctx,
+		seen: map[string]struct{}{},
+	}
+}
+
+// getSafeName returns a documentation name for an item
+// that is unique within the module.
+//
+// if the item has already been resolved by any module,
+// the previously-resolved name is returned.
+func (r *moduleConflictResolver) getSafeName(name string, item interface{}) string {
+	if safeName, ok := r.dctx.moduleConflictLinkMap[item]; ok {
+		return safeName
+	}
+
+	var prefixes []string
+	switch item.(type) {
+	case *schema.Resource:
+		prefixes = []string{"", "res-"}
+	case *schema.Function:
+		prefixes = []string{"", "fn-"}
+	case *modContext:
+		prefixes = []string{"", "mod-"}
+	default:
+		prefixes = []string{""}
+	}
+	for _, prefix := range prefixes {
+		candidate := prefix + name
+		if _, exists := r.seen[candidate]; exists {
+			continue
+		}
+		r.seen[candidate] = struct{}{}
+		r.dctx.moduleConflictLinkMap[item] = candidate
+		return candidate
+	}
+
+	glog.Error("skipping unresolvable duplicate file name: ", name)
+	return ""
+}
+
 func (mod *modContext) gen(fs codegen.Fs) error {
-	dctx := mod.docGenContext
 	modName := mod.getModuleFileName()
 
-	addFile := func(name, contents string) {
-		p := path.Join(modName, name, "_index.md")
+	conflictResolver := mod.docGenContext.newModuleConflictResolver()
+	addFile := func(safeName, contents string) {
+		p := path.Join(modName, safeName, "_index.md")
 		fs.Add(p, []byte(contents))
 	}
 
-	// Resources
-	for _, r := range mod.resources {
-		data := mod.genResource(r)
+	// Generate the index files.
+	var idxData indexData
+	{
```

Comment:
This function doesn't have much else besides what's in this block.
We don't need to nest the core logic like this.
We can declare idxData at the bottom where it's first needed.